proxy.go

```diff
@@ -230,11 +249,16 @@ func (p *Proxy) handle(ctx *Context, conn net.Conn, brw *bufio.ReadWriter) error
 		// TODO: TCPConn.WriteClose() to avoid sending an RST to the client.
 
 		return errClose
+	case req = <-reqC:
+		defer req.Body.Close()
+		break
+	case <-p.closing:
+		return errClose
 	}
-	defer req.Body.Close()
 
 	session := ctx.Session()
 
+	var err error
 	ctx, err = withSession(session)
```

Comment:
change this to 
`ctx, err := withSession(session)` 
rather than 
`var err error` above